ballerina/http_connection.bal

```diff
@@ -248,19 +259,60 @@ isolated function setPayload(anydata payload, Response response, boolean setETag
             response.setETag(payload);
         }
     } else {
-        castToJsonAndSetPayload(response, payload, "anydata to json conversion error: ", setETag);
+        processAnydata(response, payload, mediaType, setETag);
     }
 }
 
-isolated function castToJsonAndSetPayload(Response response, anydata payload, string errMsg, boolean setETag = false) {
+isolated function processAnydata(Response response, anydata payload, string? mediaType = (), boolean setETag = false) {
+    match mediaType {
+        mime:APPLICATION_FORM_URLENCODED => {
+            if payload is map<string> {
+                string|error result = retrieveUrlEncodedData(payload);
+                if result is string {
+                    response.setTextPayload(result, mime:APPLICATION_FORM_URLENCODED);
+                    if setETag {
+                        response.setETag(result);
+                    }
+                    return;
+                }
+                panic error InitializingOutboundResponseError("content encoding error: " + result.message(), result);
+            } else {
+                setJsonPayload(response, payload, setETag);
+            }
+        }
+        _ => {
+            setJsonPayload(response, payload, setETag);
+        }
+    }
+}
+
+isolated function retrieveUrlEncodedData(map<string> message) returns string|error {
+    string[] messageParams = [];
+    foreach var ['key, value] in message.entries() {
+        string encodedKey = check url:encode('key, "UTF-8");
+        string encodedValue = check url:encode(value, "UTF-8");
+        string entry = string `${'encodedKey}=${encodedValue}`;
+        messageParams.push(entry);
+    }
+    return strings:'join("&", ...messageParams);
+}
+
+isolated function setJsonPayload(Response response, anydata payload, boolean setETag) {
     var result = trap val:toJson(payload);
     if result is error {
-        panic error InitializingOutboundResponseError(errMsg + result.message(), result);
+        panic error InitializingOutboundResponseError(string `anydata to json conversion error: ${result.message()}`, result);
+    }
+    response.setJsonPayload(result);
+    if setETag {
+        response.setETag(result);
+    }
+}
+
+isolated function updateETag(Response response, anydata payload) {
```

Comment:
Do we need this function? The etag setting improvement is reverted right.

Reply:
Fixed with [c8a6e4a438948f85db3d1be5524b96f9202686be](https://github.com/ballerina-platform/module-ballerina-http/pull/830/commits/c8a6e4a438948f85db3d1be5524b96f9202686be)